### PR TITLE
Alerting: Allow Rules to Schedule to be filtered by Rule Group

### PR DESCRIPTION
**Pull request by gotjosh**
This PR enables us to filter down by rule groups when fetching alert rules. You have the ability to specify multiple groups at the same time.

### CODE REVIEW

File: pkg/services/ngalert/eval/eval.go [hunk 1]
```diff
@@ -28,7 +28,7 @@ var logger = log.New("ngalert.eval")
 type EvaluatorFactory interface {
 	// Validate validates that the condition is correct. Returns nil if the condition is correct. Otherwise, error that describes the failure
 	Validate(ctx EvaluationContext, condition models.Condition) error
-	// BuildRuleEvaluator build an evaluator pipeline ready to evaluate a rule's query
+	// Create builds an evaluator pipeline ready to evaluate a rule's query
```
Comment (gotjosh) [inline pkg/services/ngalert/eval/eval.go:31]: typo - drive-by fix.

File: pkg/services/ngalert/store/alert_rule.go [hunk 2]
```diff
@@ -397,36 +397,19 @@ func (st DBstore) GetNamespaceByUID(ctx context.Context, uid string, orgID int64
 	return folder, nil
 }
 
-func (st DBstore) getFilterByOrgsString() (string, []interface{}) {
```
Comment (gotjosh) [inline pkg/services/ngalert/store/alert_rule.go:400]: The diff on the test is also borked - but this essentially is the meat of the diff, I got rid of this function in favour of using the ORM.

Comment (yuri-tceretian) [reply pkg/services/ngalert/store/alert_rule.go:400]: I would recommend keeping this because we're migrating away from XORM to the `github.com/jmoiron/sqlx`

Comment (gotjosh) [reply pkg/services/ngalert/store/alert_rule.go:424]: The [sqlx migration has been paused](https://raintank-corp.slack.com/archives/CMSUYD6H3/p1680890445589139) for the time being - I have cleared this PR with the Grafana backend team.

**Review by JohnnyQQQQ — APPROVED** [review]
LGTM

File: pkg/services/ngalert/store/alert_rule.go [hunk 3]
```diff
@@ -466,22 +449,27 @@ func (st DBstore) GetAlertRulesForScheduling(ctx context.Context, query *ngmodel
...
-			foldersSql += " AND " + filter
-			alertRulesSql += " WHERE " + filter
+		var ids []int64
+		if len(st.Cfg.DisabledOrgs) > 0 {
+			for orgID := range st.Cfg.DisabledOrgs {
+				ids = append(ids, orgID)
+			}
+		}
+
+		alertRulesSql := sess.Table("alert_rule").Select("*")
```
[the remark this answers is not in the document]
Comment (gotjosh) [reply pkg/services/ngalert/store/alert_rule.go:459]: Do you know something I don't about `AllCols`? 😆 It seems to be a update use only thing. 

```
// AllCols update use only: update all columns
func (statement *Statement) AllCols() *Statement {
	statement.useAllCols = true
	return statement
}
```

While I do see that `Table` and `Select` return the same type -- I couldn't find where is it implicit. I ran the tests without the select and they pass. I guess that's good enough?

File: pkg/services/ngalert/store/alert_rule.go [hunk 4]
```diff
@@ -466,22 +449,27 @@ func (st DBstore) GetAlertRulesForScheduling(ctx context.Context, query *ngmodel
...
 
 		rule := new(ngmodels.AlertRule)
-		rows, err := sess.SQL(alertRulesSql, args...).Rows(rule)
+		rows, err := alertRulesSql.Rows(rule)
 		if err != nil {
 			return fmt.Errorf("failed to fetch alert rules: %w", err)
 		}
-		defer func() {
-			_ = rows.Close()
-		}()
```
[the remark this answers is not in the document]
Comment (gotjosh) [reply pkg/services/ngalert/store/alert_rule.go:484]: Good catch - I think I dropped it in the new rebase.

File: pkg/services/ngalert/store/alert_rule.go [hunk 5]
```diff
@@ -466,22 +449,27 @@ func (st DBstore) GetAlertRulesForScheduling(ctx context.Context, query *ngmodel
...
 	return st.SQLStore.WithDbSession(ctx, func(sess *db.Session) error {
-		foldersSql := "SELECT D.uid, D.title FROM dashboard AS D WHERE is_folder IS TRUE AND EXISTS (SELECT 1 FROM alert_rule AS A WHERE D.uid = A.namespace_uid)"
-		alertRulesSql := "SELECT * FROM alert_rule"
-		filter, args := st.getFilterByOrgsString()
-		if filter != "" {
-			foldersSql += " AND " + filter
-			alertRulesSql += " WHERE " + filter
+		var ids []int64
+		if len(st.Cfg.DisabledOrgs) > 0 {
+			for orgID := range st.Cfg.DisabledOrgs {
```
[the remark this answers is not in the document]
Comment (gotjosh) [reply pkg/services/ngalert/store/alert_rule.go]: good shout.

File: pkg/services/ngalert/store/alert_rule.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (gotjosh) [reply pkg/services/ngalert/store/alert_rule.go:457]: I'm happy with this -- let's roll with it.

File: pkg/services/ngalert/store/alert_rule.go [hunk 6]
```diff
@@ -495,8 +483,16 @@ func (st DBstore) GetAlertRulesForScheduling(ctx context.Context, query *ngmodel
 		}
 
 		query.ResultRules = rules
+
 		if query.PopulateFolders {
-			if err := sess.SQL(foldersSql, args...).Find(&folders); err != nil {
+			foldersSql := sess.Table("dashboard").Alias("d").Select("d.uid, d.title").
+				Where("is_folder = ?", true).
```
[the remark this answers is not in the document]
Comment (gotjosh) [reply pkg/services/ngalert/store/alert_rule.go]: I found `d.store.GetDialect().BooleanStr(true)` used in other places.

There's also `	err := m.sess.Table(&dashboard{}).Where("is_folder=?", true).Distinct("org_id").Find(&rows)` which is part of `getOrgsIDThatHaveFolders` that is function used as part of the migration.

I went with the former so I think we're safe.

**Review by alexweav — APPROVED** [review]
LGTM
